models/models.js

```diff
@@ -1,65 +1,132 @@
+      }
 
-exports.fetchTopics = () => {
-    return db.query(
-        `SELECT * FROM topics`
-    )
-    .then(({rows}) => {
-        return rows
+      return result.rows[0];
     })
 };
-//future challenge here is util function that can be reused => to incorporate in other funcs
-exports.checkArticleExists = async(article_id) => {
-    const dbOutput = await db
-    .query(
-        `SELECT * FROM articles WHERE article_id = $1;`, [article_id]
-        ); 
-    if (dbOutput.rows.length === 0) {
-        return Promise.reject({ status: 404,  message: "item does not exist"});
-    }
-    return dbOutput.rows[0];
-}
- 
-exports.fetchArticleById = (id) => {
-    return db
-    .query( `SELECT * FROM articles WHERE article_id = $1;`, [id])
+module.exports.checkArticleExists = checkArticleExists;
+
+checkUsernameExists = async (username) => {
+  return db.query(
+    "SELECT username FROM users WHERE username = $1;", [username])
     .then((result) => {
-   
-        if (result.rows.length === 0) {
-            return Promise.reject({ status: 404,  message: "item does not exist"});
-         }
 
-        return result.rows[0];
+        if(result.rows.length === 0) {
+            return Promise.reject({ status: 404, message: 'Username does not exist'});
+        }
+      return result.rows[0];
     })
-}
+};
+module.exports.checkUsernameExists = checkUsernameExists;
+
+exports.fetchArticleById = (id) => {
+  return db
+    .query(`SELECT * FROM articles WHERE article_id = $1;`, [id])
+    .then((result) => {
+      if (result.rows.length === 0) {
+        return Promise.reject({ status: 404, message: "item does not exist" });
+      }
+
+      return result.rows[0];
+    });
+};
 
 //5
 exports.fetchArticles = () => {
-    const articlesQuery = `
+  const articlesQuery = `
     SELECT a.article_id, a.title, a.author, a.topic, a.created_at, a.votes, a.article_img_url, COUNT(c.comment_id) AS comment_count 
 FROM articles AS a 
 LEFT JOIN comments AS c 
 ON c.article_id = a.article_id
 GROUP BY a.article_id
 ORDER BY a.created_at DESC; 
 `;
-    return db.query(articlesQuery)
-        .then(({rows}) => {
-           return rows
-    })
-}
+  return db.query(articlesQuery).then(({ rows }) => {
+    return rows;
+  });
+};
 //6
 exports.selectArticleComments = (articleId) => {
-    return db.query(`
+  return db
+    .query(
+      `
     SELECT * FROM comments WHERE article_id = $1 ORDER BY created_at DESC; 
-    `,  [articleId])
+    `,
+      [articleId]
+    )
     .then((result) => {
-        return result.rows;
+      return result.rows;
+    });
+};
+//7
+exports.insertComment = (newComment, id) => {
+  const { username, body } = newComment;
+
+  return Promise.all([
+    checkUsernameExists(newComment.username),
+    checkArticleExists(id)
+  ])
+    .then(([usernameExists, articleExists]) => {
+      if (!usernameExists) {
+        return Promise.reject({ status: 404, message:'Username does not exist' });
```

Comment:
It shouldn't be necessary to check the username exists for POST, it is referenced in the users table, so PSQL will throw it's own error for you to catch.

Reply:
Ali told me to do that test so i added functionality and test

Reply:
<img width="897" alt="Screenshot 2023-10-05 at 08 55 51" src="https://github.com/annc4st/be-nc-news-project/assets/106443136/beb330d7-25e4-4f6a-a05b-77d1c4fa507c">

Reply:
Ali told you _not_ to do it 👍